### PR TITLE
Try: Add a clickable Group setup state

### LINKED ISSUE #33025: Default Appender: Add select-state before showing inserter
## What problem does this address?

The default appender component makes it easy to insert a block in its place. But it makes it hard to select the block itself:

problem

Additionally, since this particular appender serves as a de-facto placeholder/setup state for some blocks (shown here, empty column and group blocks), the single bare plus icon is perhaps insufficient indication of what you're looking at.

#32108 discusses the same problem and how it affects the Template Part block, and a solution being discussed [in the comments](https://github.com/WordPress/gutenberg/issues/32108#issuecomment-849475428): build in a selection state before the insertion stage (aka. "clickthrough"):

- The resting state shows the type of block and a clickable area
- The selected state surfaces the inserter button

## What is your proposed solution?

Today, preinserted groups and columns look like this:

Setup states   appenders

By adding a selection state first, that resting can state serve both as block indication and clickable area. A click here would surface the inserter:

Setup states   appenders after

Here's a basic prototype:

prototype

This **select the block first, then insert** behavior would fix the issue described in #32108. #30116 and #30115 are adjacent, but it's not clear if this pattern will work in those cases.

### PR DESCRIPTION
**Pull request by jasmussen**
## What?

Fixes #33025. 

This is a pretty experimental PR, I'd love thoughts on whether this is fundamentally a good idea.

If you insert an empty group block, there's a big appender button to add content inside. That's good enough, as your next step is likely going to be filling in the group. The downside is that you can't, in the canvas, select an empty group to delete it, because you always click the inserter button when trying to select the empty group, as shown in this gif:

group before

This PR follows in the footsteps of changes to the Row and Stack variations, and lets the block have a resting state showing a dashed outline — a literal placeholder. On select, the appender button becomes visible. In effect this means you can select and delete the block, or you can select the block and then click again to insert content inside, as shown in this gif:

group after

## Why?

The ability to select the row and stack blocks without invoking the inserter popout is a nice aspect that I find myself missing on the base Group block.

## Testing Instructions

Insert an empty group block. Deselect it, then select it again. 

Be sure to test drag and drop, uploads, and various other container blocks. These changes should not affect any of those, and should only change the appearance of the empty group.

### CODE REVIEW

**Comment by ZebulanStanphill** [comment]
Another benefit of adding this extra click is that it gives you a chance to see exactly what block you'll be inserting into. Before, there was no easy way to know that you were inserting into a Group block versus some other container block e.g. a Column.

**Comment by youknowriad** [comment]
I like this but I also wonder whether using the "block overlay" (when the group is empty) could be a good way to implement this without adding a specific solution to the group block. It's something we'd be able to leverage pretty easily after we land #40649

**Comment by jameskoster** [comment]
Happy to see a solution to this one! Should we apply the same treatment to empty columns? It always bugs me to see the [+] when using columns for layout.

Tangentially, is it _necessary_ to simulate multiple blocks in empty rows/stacks? I'm wondering if all empty groups could have the same appearance.

> without adding a specific solution to the group block

The dashed outline isn't unique, we use it for other placeholders like Featured Image and Site Logo.

**Comment by jasmussen** [comment]
> The dashed outline isn't unique, we use it for other placeholders like Featured Image and Site Logo.

Agreed, and now for Navigation as well. But if I read Riad's comment right, what we need is a more generic solution, rather than one that requires custom CSS on a per-block basis.

In this case, I don't know how to best do that, though: and the Row/Stack blocks have some bespoke CSS to make this happen as well for the emulated row and stack appearances. So I guess the question is, how generic can we be without those blocks losing their setup-state appearance?

**Comment by jasmussen** [comment]
> Happy to see a solution to this one! Should we apply the same treatment to empty columns? It always bugs me to see the [+] when using columns for layout.

If we could make it generic as suggested above, we might get this for free. Then the row and stack blocks would be the exceptions to the "generic" rule 🤔

**Comment by youknowriad** [comment]
> Agreed, and now for Navigation as well. But if I read Riad's comment right, what we need is a more generic solution, rather than one that requires custom CSS on a per-block basis.

Yes, that's exactly my suggestion :) Basically we could just pass `isDisabled` (see other PR) and add a classname (overlay) to useBlockProps and it should work. I still think we can simplify even more and just have a `showOverlay` that does both: disable and add the class.

Basically my suggestion is the following: When the group blocks are empty, enable the overlay mode (like template parts and reusable blocks).

**Comment by jameskoster** [comment]
Do we definitely want to retain overlay mode? (context: https://github.com/WordPress/gutenberg/issues/32163#issuecomment-1110859404)

**Comment by jasmussen** [comment]
Coming back to this one, I've rebased it and fixed an issue with a classname change. As a reminder, here's before:
before

Here's after:

after

It's not a ton of CSS, and it arguably improves the empty setup state of the group. Should we land it? We can always tear out the CSS again if/when we're able to look at a more generic overlay state?

**Review by jameskoster — APPROVED** [review]
This is working as advertised for me, and feels like a nice quality of life improvement.

**Comment by jasmussen** [comment]
Thank you. Looks like the test failure is legitimate, so I'll have to look into that.

**Comment by jasmussen** [comment]
This PR, as it turns out, is slightly blocked on #43115 which is a bit more urgent.

**Comment by jasmussen** [comment]
This will need a rebase after #43169 is landed.

**Comment by jasmussen** [comment]
This one is ready for another view. It's rebased per recent changes, and it fixes an issue with nested empty groups or rows.

Here's some test content:

```
<!-- wp:paragraph -->
<p>Stack:</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"flex","orientation":"vertical"}} -->
<div class="wp-block-group"></div>
<!-- /wp:group -->

<!-- wp:paragraph -->
<p>Row:</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"flex","flexWrap":"nowrap"}} -->
<div class="wp-block-group"></div>
<!-- /wp:group -->

<!-- wp:paragraph -->
<p>A row inside a row:</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"flex","flexWrap":"nowrap"}} -->
<div class="wp-block-group"><!-- wp:group {"layout":{"type":"flex","flexWrap":"nowrap"}} -->
<div class="wp-block-group"></div>
<!-- /wp:group --></div>
<!-- /wp:group -->

<!-- wp:paragraph -->
<p>A single Group:</p>
<!-- /wp:paragraph -->

<!-- wp:group -->
<div class="wp-block-group"></div>
<!-- /wp:group -->

<!-- wp:paragraph -->
<p>A group inside a group:</p>
<!-- /wp:paragraph -->

<!-- wp:group -->
<div class="wp-block-group"><!-- wp:group -->
<div class="wp-block-group"></div>
<!-- /wp:group --></div>
<!-- /wp:group -->
```

**Comment by jameskoster** [comment]
Still working well for me :) 

Seeing as the placeholder style is beginning to pop up in a few places, should we create a mixin so that it's easier to update?

**Comment by jasmussen** [comment]
> should we create a mixin so that it's easier to update?

We should do something, yes. Mixin may be the most straight-forward, I can try that next week. I was hoping for some of the placeholder stuff we could componentize, but we may need both 🤔 

Thanks for testing!

**Pull request merged**
